src/github.com/getlantern/detour/ioloop.go

```diff
@@ -0,0 +1,348 @@
+package detour
+
+import (
+	"bytes"
+	"io"
+	"sync/atomic"
+)
+
+// to pass result of read operation to merger from underlie connection
+type innerReadResult struct {
+	// the connection which performs this read
+	c conn
+	// buffer holding received data, it's not necessarily the same
+	// buffer as innerReadRequest.
+	buf []byte
+	// IO error, if any
+	err error
+}
+
+type innerReadRequest struct {
+	buf      []byte
+	chResult chan innerReadResult
+}
+
+// ioLoop is the core of detour. It waits for connections and handles
+// read/write requests
+func (dc *Conn) ioLoop() {
```

Comment:
Because reads and writes are handled by the same goroutine, this ends up being half-duplex.  For TLS negotation and typical HTTP traffic, that's not a problem, but if someone is proxying stuff through lantern that benefits from full duplex, this might be an issue.